lib/sessions.js

```diff
@@ -98,6 +135,167 @@ class ClientSession extends EventEmitter {
+  }
+
+  /**
+   * Commits the currently active transaction in this session.
+   *
+   * @param {Function} [callback] optional callback for completion of this operation
+   * @return {Promise} A promise is returned if no callback is provided
+   */
+  commitTransaction(callback) {
+    if (typeof callback === 'function') {
+      endTransaction(this, 'commitTransaction', callback);
+      return;
+    }
+
+    return new Promise((resolve, reject) => {
+      endTransaction(
+        this,
+        'commitTransaction',
+        (err, reply) => (err ? reject(err) : resolve(reply))
+      );
+    });
+  }
+
+  /**
+   * Aborts the currently active transaction in this session.
+   *
+   * @param {Function} [callback] optional callback for completion of this operation
+   * @return {Promise} A promise is returned if no callback is provided
+   */
+  abortTransaction(callback) {
+    if (typeof callback === 'function') {
+      endTransaction(this, 'abortTransaction', callback);
+      return;
+    }
+
+    return new Promise((resolve, reject) => {
+      endTransaction(
+        this,
+        'abortTransaction',
+        (err, reply) => (err ? reject(err) : resolve(reply))
+      );
+    });
+  }
+}
+
+// see: https://github.com/mongodb/specifications/blob/master/source/retryable-writes/retryable-writes.rst#terms
+const RETRYABLE_ERROR_CODES = new Set([
+  6, // HostUnreachable
+  7, // HostNotFound
+  64, // WriteConcernFailed
+  89, // NetworkTimeout
+  91, // ShutdownInProgress
+  189, // PrimarySteppedDown
+  9001, // SocketException
+  11600, // InterruptedAtShutdown
+  11602, // InterruptedDueToReplStateChange
+  10107, // NotMaster
+  13435, // NotMasterNoSlaveOk
+  13436 // NotMasterOrSecondary
+]);
+
+function isRetryableError(error) {
+  if (
+    RETRYABLE_ERROR_CODES.has(error.code) ||
+    error instanceof MongoNetworkError ||
+    error.message.match(/not master/) ||
+    error.message.match(/node is recovering/)
+  ) {
+    return true;
+  }
+
+  return false;
+}
+
+function resetTransactionState(clientSession) {
+  clientSession.transactionOptions = null;
+}
+
+function endTransaction(clientSession, commandName, callback) {
+  if (!assertAlive(clientSession, callback)) {
+    // checking result in case callback was called
+    return;
+  }
+
+  if (!clientSession.inTransaction()) {
+    if (clientSession.autoStartTransaction) {
+      clientSession.startTransaction();
+    } else {
+      callback(new MongoError('No transaction started'));
+      return;
+    }
+  }
+
+  if (clientSession.serverSession.stmtId === 0) {
+    // The server transaction was never started.
+    resetTransactionState(clientSession);
+    callback(null, null);
+    return;
+  }
+
+  const command = { [commandName]: 1 };
+  if (clientSession.transactionOptions.writeConcern) {
+    Object.assign(command, { writeConcern: clientSession.transactionOptions.writeConcern });
+  } else if (clientSession.clientOptions.w) {
```

Comment:
as far as I can tell, we don't have a guarantee that `clientSession.clientOptions != null`, so we might want to guard this.